infrastructure/src/main/java/org/corfudb/infrastructure/RemoteMonitoringService.java

```diff
@@ -234,19 +236,25 @@ private void runHealingDetectorTask() {
             final PollReport pollReport = healingDetector.poll(layout, corfuRuntime);
 
             clusterStateContext.updateResponsiveNodes(pollReport.getChangedNodes());
-            pollReport.getClusterStateMap().forEach((s, clusterStatus)
-                    -> clusterStateContext.updateNodeState(clusterStatus, layout));
+            pollReport.getClusterStateMap().forEach((s, clusterStatus) ->
+                    clusterStateContext.updateNodeState(clusterStatus, layout));
+
+            Set<String> healedNodes = new TreeSet<>(recommendationEngine
+                    .healedServers(clusterStateContext.getClusterState(), layout));
 
-            if (pollReport.isChangedNodesEmpty()) {
+            if (healedNodes.isEmpty()) {
```

Comment:
failed node handling already taking unfilled slots in layout into account. Wouldn't we want to have the same check through healing as well? It might improve the time to discover the unfilled slots.

Reply:
The unfilled slots check is not enough. You would also need to correct out of phase epochs. So duplicating all the logic would be cumbersome.